### ISSUE TITLE
array runtime type not considered in == operator on types

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#15721]
``` chapel
var A:[1..10] int;
var B:[1..20] int;

writeln(A.type == B.type);
```

I would expect this program to print out false, because A's type includes the domain `{1..10}` but B's type includes the different domain `{1..20}`.

**Comment by bradcray** [comment]
For me, this issue/question gets back to the longstanding discussion about whether we should introduce `.stype` and `.dtype` queries to distinguish between whether we're talking about the static or dynamic aspects of an expression's type and what the default of `.type` should be.  I think that, currently, we unfortunately sometimes (most of the time?) treat `.type` as the static type but occasionally treat it as the dynamic type (?).

**Comment by bradcray** [comment]
Though maybe I'm wrong about that, since in this case, it's the identity of the domain that differs rather than the domain's value?  Can we always know whether two arrays share the same domain or not statically?

**Comment by mppf** [comment]
I think an alternative way to phrase what you are getting at is this.

Do we want `someVar.type == someOtherVar.type` to consider runtime types? 

The answer is not necessarily "yes" but I thought it would be when I created the issue. I think you're saying that with `.stype` and `.dtype` the programmer could be explicit about whether or not they wished the runtime type to be considered.

In favor of considering runtime types there:
 * More consistent

In favor of not considering runtime types there:
 * Doing so might mess up param folding in some common case (maybe?)

**Comment by mppf** [comment]
>  Can we always know whether two arrays share the same domain or not statically?

I don't think so; e.g. imagine having an array of domains `MyDomains` and then initializing and returning an array declared over the domain `MyDomains[i]` where `i` is a `config const`, say.

However we might be able to specify the cases in which the compiler can know that two arrays share the same domain. (It might also happen in other cases, but the compiler wouldn't know it).

**Comment by bradcray** [comment]
I ran back into this issue today while working on implementing `==` and `!=` for records in module code in #28731 .  Specifically, Daniel pointed out that if two records had a type field with the same static type, the dynamic types of the fields might differ.  That led me to think that maybe the module-based `==`/`!=` on records should compare type fields, but since they only look at static types today, that wasn't easy to do.

[edit:]

One thing I wonder about is whether having `==` always consider static and dynamic information for types is correct/appropriate or whether we'd want distinct ways of comparing static/dynamic type information.  For example, `where` clauses often use `==` on types, and need to be compile-time evaluated.  So we could either have the compiler automatically change `==` on types into "static-only" vs. "static & dynamic" comparisons (based on context?), or else we'd need two different ways to compare them.
